onnx/optimizer/passes/shape_inference.h

```diff
@@ -0,0 +1,46 @@
+#ifndef SHAPE_INFERENCE_H
+#define SHAPE_INFERENCE_H
+
+#include "onnx/defs/schema.h"
+#include "onnx/optimizer/passes/optimize_pass.h"
+
+namespace ONNX_NAMESPACE { namespace optimization {
+
+// Shape inference is best-effort, and there is no guarantee of which
+// Values in the Graph will end up with full shape and type inference.
+struct ShapeInference : public OptimizePass {
```

Comment:
nit: Making classes and structs `final` makes sure your class doesn't fall in one of the many inheritance traps (say not having virtual destructors) because someone inherited from it and you didn't anticipated it.
Also, making it `final` allows the compiler to better optimize some code like type casts.

In this particular case, it doesn't apply because the destructor of `OptimizePass` is already virtual, but it's still a good general C++ rule, and the compiler optimization point still applies.